Source/WebCore/platform/graphics/gstreamer/GStreamerSinksWorkarounds.cpp

```diff
@@ -0,0 +1,331 @@
+        if (mode == WorkaroundMode::ForceEnable) {
+            GST_DEBUG("BaseSinkPositionFlushWorkaroundProbe: forcing workaround to be enabled.");
+            return true;
+        }
+        if (mode == WorkaroundMode::ForceDisable) {
+            GST_DEBUG("BaseSinkPositionFlushWorkaroundProbe: forcing workaround to be disabled.");
+            return false;
+        }
+
+        return !webkitGstCheckVersion(1, 24, 0);
+    }
+
+    static void initializeIsNeeded()
+    {
+        s_isNeeded = checkIsNeeded();
+        GST_DEBUG("BaseSinkPositionFlushWorkaroundProbe is%s needed in this system.", s_isNeeded ? "" : " NOT");
+    }
+
+    static bool flushIsNonResetting(GstEvent* event)
+    {
+        gboolean resetTime = TRUE;
+        gst_event_parse_flush_stop(event, &resetTime);
+        return !resetTime;
+    }
+
+    static GstPadProbeReturn probe(GstPad* pad, GstPadProbeInfo* info, void* userData)
+    {
+        auto* self = static_cast<BaseSinkPositionFlushWorkaroundProbe*>(userData);
+        GstBaseSink* basesink = GST_BASE_SINK(GST_PAD_PARENT(pad));
+        GST_TRACE_OBJECT(pad, "m_handlingFlushStop: %s, pad is flushing: %s, received: %" GST_PTR_FORMAT, boolForPrinting(self->m_handlingFlushStop), boolForPrinting(GST_PAD_IS_FLUSHING(pad)), info->data);
+        if (!self->m_handlingFlushStop && (info->type & GST_PAD_PROBE_TYPE_EVENT_FLUSH) && GST_EVENT_TYPE(info->data) == GST_EVENT_FLUSH_STOP
+            && flushIsNonResetting(GST_EVENT(info->data)) && basesink->segment.format != GST_FORMAT_UNDEFINED)
+        {
+            GST_DEBUG_OBJECT(pad, "Received non-resetting FLUSH_STOP while we have a valid segment... propagating the FLUSH_STOP.");
+
+            self->m_handlingFlushStop = true;
+            gst_pad_send_event(pad, GST_EVENT(info->data));
+            self->m_handlingFlushStop = false;
+
+            // For non-resetting FLUSH_STOP events basesink preserves the segment but still sets have_newsegment to FALSE.
+            // Until the fix was introduced, this makes gst_base_sink_get_position() fail.
+            // The workaround consists on setting have_newsegment to TRUE in that case after the FLUSH_STOP is handled
+            // internally by basesink.
+            //
+            // This is a reasonably safe workaround, as the only other place where have_newsegment is read is in a warning
+            // message in gst_base_sink_chain_unlocked() for when a buffer is pushed without a segment, something that
+            // shouldn't happen in the first place.
+            GST_DEBUG_OBJECT(pad, "Non-resetting FLUSH_STOP has been handled by this element and its downstream. Now force position queries to work by setting basesink->have_newsegment to TRUE.");
+            basesink->have_newsegment = TRUE;
+
+            return GST_PAD_PROBE_HANDLED;
```

Comment:
Where did you see this? gst_pad_send_event() takes `[transfer full]` so we shouldn't be leaking...

Reply:
You're right, there is no leak because of that transfer full that I missed. But not that you mention it, it would be better to use `GST_PAD_PROBE_INFO_EVENT` instead of `GST_EVENT(info->data)`